### DISCUSSION TITLE
How do BPM changes work?

### DISCUSSION THREAD

**Discussion by amyspark-ng** [FunkinCrew/Funkin#3887] [Q&A] [answered]
Would someone so kindly explain to me how the code for BPM changes work in the conduction?

I know they come from an array that is on the metadata file, and my issue is that they all have a step time, how do i exactly find out this, the conversion for the time and the step time considering the bpm changes, is confusing!!!

**Comment by KoloInDaCrib** [comment]
Basically in every BPM change there is a timestamp in milliseconds as to when it happens (with an alias "t"). Upon calling the function [`mapTimeChanges`](https://github.com/FunkinCrew/Funkin/blob/9e13ea0b2c0799d9a7fae7db56eddb05898739d0/source/funkin/Conductor.hx#L521), it fills the array `timeChanges` with the BPM changing data, sorted by their timestamp from lowest to highest. You'd have to call mapTimeChanges manually after loading a song with anything other than FunkinSound's playMusic function.

After that in each [`update`](https://github.com/FunkinCrew/Funkin/blob/9e13ea0b2c0799d9a7fae7db56eddb05898739d0/source/funkin/Conductor.hx#L405) call of the Conductor it checks for what timeChange to currently use based on the timestamp with this piece of code
![image](https://github.com/user-attachments/assets/ddaaf4e8-a24c-4187-b0ca-dbbddd0d64ae)
(side note I feel like it would've been somewhat easier with a decreasing array and a while loop)

As the last step it just sets the times of how long a measure/beat/step takes based on this calculation. Hope this helps yeah !!!

**Comment by amyspark-ng** [comment]
thank you so much for the reply!!

Up until then i understand, my issue is:

1. how does the game calculate the amount of total steps in the song when the interval is not consistent
2. also the current time in beats and steps
3. also the fact that the events always have a property like beatTime and stepTime, these are necessary for calculating steps and beats in the song, but how did they put step time in the events when the time to step function requires step time of the time change?!

**Reply by KoloInDaCrib** [reply, marked as answer]
1. There is a neat little function in Conductor called `getTimeInSteps` which returns how many steps there are in the given parameter representing the time in milliseconds. My guess for calculating the total amount of steps would be just using `FlxG.sound.music.length` as the function parameter lol

2. Current Time in Steps is calculated as such (according to the constructor's update function)
- Multiply the beatTime of the last change with 4 (default amount for steps per beat)
- Take the difference of the song's current position with time change's timestamp and divide it by how much a step lasts in milliseconds (60 divided by current BPM multiplied by 1000 for the ms and then divided by timeSignatureNumerator)
Current Time in Beats is literally just Current Time in Steps but divided by 4 - incredible

3. mapTimeChanges function sets the beatTime of the bpm changes according to this part so that no value is null and everything works as intended
![image](https://github.com/user-attachments/assets/4b7c7530-6bf8-483a-b0ff-fbdda7d3ecfa)

**Reply by amyspark-ng** [reply]
woah feels like i just discovered a whole part of conductor code i never seen! im gonna mark this as an answer, since it can't really test it right now

i was asking because i am making a rhythm game and was interested in adding bpm changes, but thank you so much for the help! hopefully it'll work for me

**Reply by KoloInDaCrib** [reply]
No problem! Good luck on your rhythm game project!!
